src/modules/stripe/stripe.service.ts

```diff
@@ -0,0 +1,159 @@
+    payment: PaymentDto,
+  ): Promise<Stripe.Response<Stripe.Checkout.Session>> {
+    const { productIds, total, shippingPrice } = payment;
+
+    try {
+      const user = await this.usersServise.getUserById(customerId);
+
+      if (!user) {
+        throw new NotFoundException(Errors.USER_NOT_FOUND);
+      }
+
+      const products = await this.productsService.findByIds(productIds);
+
+      if (products.length < productIds.length) {
+        throw new NotFoundException(Errors.SOME_PRODUCTS_NOT_FOUND);
+      }
+
+      const productsTotalPrice = products.reduce(
+        (sum, product) => sum + Number(product.price),
+        0,
+      );
+
+      if (
+        Math.abs(productsTotalPrice + shippingPrice - total) > minDifference
+      ) {
+        throw new ConflictException(Errors.PRICES_DO_NOT_MATCH);
+      }
+
+      const lineItems: Stripe.Checkout.SessionCreateParams.LineItem[] =
+        products.map((product) => {
+          return {
+            price_data: {
+              currency: 'usd',
+              product_data: {
+                name: product.name,
+                images: [product.images[0]],
+              },
+              unit_amount: Number(product.price) * centsInDollar,
+            },
+            quantity: 1,
+          };
+        });
+
+      if (shippingPrice > 0) {
+        lineItems.push({
+          price_data: {
+            currency: 'usd',
+            product_data: {
+              name: 'Shipping',
+              images: [CANADA_POST_LOGO],
+            },
+            unit_amount: shippingPrice * centsInDollar,
+          },
+          quantity: 1,
+        });
+      }
+
+      const session = await this.StripeApi.checkout.sessions.create({
+        line_items: lineItems,
+
+        payment_intent_data: {
+          transfer_group: mockOrders[0].transferGroup,
+        },
+        mode: 'payment',
+        success_url: this.configService.get<string>('STRIPE_SUCCESS_URL'),
+        cancel_url: this.configService.get<string>('STRIPE_CANCEL_URL'),
+      });
+
+      return session;
+    } catch (error) {
+      if (
+        error instanceof NotFoundException ||
+        error instanceof ConflictException
+      ) {
+        throw error;
+      }
+      throw new InternalServerErrorException(Errors.PAYMENT_ERROR);
+    }
+  }
+
+  async webhookHandler(event: Stripe.Event): Promise<{ received: boolean }> {
+    if (event.type === 'checkout.session.completed') {
+      const session = event.data.object as Stripe.Checkout.Session;
+      const paymentIntentId = session.payment_intent as string;
+
+      const paymentIntent =
+        await this.StripeApi.paymentIntents.retrieve(paymentIntentId);
+      const transferGroup = paymentIntent.transfer_group;
+
+      const userId = mockOrders.find(
+        (order) => order.transferGroup === transferGroup,
+      ).userId;
+
+      await this.cartService.emptyCart(userId);
+    }
+
+    return { received: true };
+  }
+
+  async checkSignature(
```

Comment:
are you sure that it should be open route

Reply:
The webhook route is used by Stripe to transmit transaction data. It is public, but the data is verified using a secret key.